qiime2/plugin/plugin.py

```diff
@@ -201,9 +205,14 @@ def register_semantic_type_to_format(self, semantic_type, artifact_format):
                 raise ValueError("%r has a predicate, differentiating format"
                                  " on predicate is not supported.")
 
-        self.type_formats.append(TypeFormatRecord(
+        record = TypeFormatRecord(
             type_expression=semantic_type,
-            format=artifact_format, plugin=self))
+            format=artifact_format, plugin=self)
+
+        self.type_formats.append(record)
+        for type_ in record.type_expression:
```

Comment:
When we iterate over record.type_expression, we are stepping through a "semantic type". Does this mean each iteration is a type fragment? If so, is it appropriate to add these to `self.types` instead of `self.type_fragments`? 

If we're stepping over the `semantic_type` itself, is iteration necessary at all?

Reply:
They are being added to self.types, for each fragment in the type_expression we construct a `SemanticTypeRecord` of the full type. . . Or something like that. I'm a bit fuzzy on it honestly, but that seems to be what's happening and is definitely the intent.

Reply:
I must be misunderstanding something here. If each `type_` we iterate over is a fragment, won't passing `type_` to `SemanticTypeRecord()` create records of type _fragments_, which we are then putting in `self.types`?

Reply:
That's the part that I don't really understand

Reply:
Correction, now that I actually look at where this actually is, `TypeFormatRecord` associates a full type with its canonical format. So `type_` is a full type expression at this point.

Reply:
Sorry if I'm completely missing the boat here. Is it possible for `record.type_expression` to hold more than one "full type expression"? If not, why are we iterating over it?

Reply:
> Is it possible for `record.type_expression` to hold more than one "full type expression"?

Yes. For example, this type expression can be [found in the wild](https://github.com/qiime2/q2-types/blob/22aa622f311881459373b9286b6adf526d7ef5df/q2_types/feature_table/_type.py#L38-L43): `FeatureTable[Frequency | RelativeFrequency | PresenceAbsence | Balance | Composition | PercentileNormalized]`

A MWE:

```python
import qiime2
from q2_types.feature_table import FeatureTable, Frequency, RelativeFrequency, PresenceAbsence

exp = FeatureTable[Frequency | RelativeFrequency | PresenceAbsence]
for type_ in exp:
    print(type_)
```

```
FeatureTable[Frequency]
FeatureTable[RelativeFrequency]
FeatureTable[PresenceAbsence]
```

Also, I just want to point out, @Oddant1 didn't add this loop, he moved it from the `PluginManager` (not that it really matters, I just want to clarify that this code didn't appear out of thin air).

Reply:
Thanks for the example, @thermokarst!